feat(api): Consistently allow specifying namespaces and versions of adapters and lids

SyntaxColoring opened the pull request:
## Overview

Closes the Python part of EXEC-1750.

The Python Protocol API has a number of "combination load methods" that let you load multiple labware in a single call, like a main labware plus an adapter, or a main labware plus a lid.

* `Labware`
  * `load_labware()`
* `ModuleContext`
  * `load_labware()`
  * `set_stored_labware()`
* `ProtocolContext`
  * `load_labware()`
  * `load_lid_stack()`

There are two things about these APIs today that are problematic:

1. There are no parameters to let you manually specify the namespace and version of the adapter/lid, like you can for the main labware.

2. Without those parameters, the Python Protocol API tries to automatically pick the namespace and version values for the adapter/lid. The specific way that it does this is problematic. It will sometimes try to match the the version of the adapter/lid to the version of the main labware. But there's no reason to believe that that version of the adapter/lid will even exist. For example, you could have a well plate that goes up to v5 and a lid that only goes up to v2.

## Changelog

For all of the "combined load methods":

* Add new parameters:
  * `lid_namespace`
  * `lid_version`
  * `adapter_namespace`
  * `adapter_version`
* In the case where these parameters are not supplied, stop copying their values from the namespace and version of the "main" labware. Instead:`namespace` defaults to searching `opentrons` and custom labware, and `version` defaults to something `apiLevel`-dependent. This is consistent with how the basic case of `ProtocolContext.load_labware(load_name, location)` works.
* Gate this behind ~~`"apiLevel": "2.25"`, which is the `apiLevel` being released in robot software v8.6.0.~~, `"apiLevel": "2.26"`, which is the next one in line to be released.
* Update docstrings enough to be correct, though perhaps not enough to be good. I'll leave refinements to follow-up PRs.

## Test Plan and Hands on Testing

* Try the snippet in EXEC-1750.

## Review requests

* **Is this how we want to resolve EXEC-1750?**
* ~~**Do we want this to target chore_release-8.6.0?**~~
  * Decision with @sfoster1: No, too late for 8.6.0. ~~Merge it into `edge`. Perhaps later we will do a v8.6.1, in which case we will cherry-pick this into that branch.~~
  * Per @ddcc4: Merge it into `hold-for-8.6.1`.
* **Make sure I didn't forget any methods.** I'm not really familiar with the whole zoo anymore. I basically just grepped for `lid:` and `adapter:`, so we should make sure that that didn't miss anything.
* **Make sure I didn't make any silly logic mistakes like flipping less-than vs. greater-than.**

## Risk assessment

Medium. The actual changes are as easy as they come, but this is a rushed change ~~targeting a release branch after a feature freeze, so we should be wary of it. This PR only exists because I think the other options are more risky in their own ways.~~

SyntaxColoring committed:
Add (lid|adapter)_(namespace|version) args everywhere there's (lid|adapter) arg.

SyntaxColoring committed:
The traditional utterances

SyntaxColoring committed:
Oops, missed one.

SyntaxColoring committed:
Update test for set_stored_labware().

ddcc4 commented:
This would have to go into `chore_release-8.6.0`, right?

ecormany reviewed (commented):
We can add docstrings here or make this an implementation-only PR. I've made a [ticket](https://opentrons.atlassian.net/browse/RTC-820), so it won't get lost if we omit them here.

SyntaxColoring committed:
More unit test updates.

SyntaxColoring committed:
Comment.

SyntaxColoring committed:
Docstrings.

sfoster1 reviewed (changes requested):
This looks good, but can we bounce the API version to a variable and then also stringify the variable in the `until_version` fields of the version variables? That way there's fewer things to remember to change

ecormany commented on api/src/opentrons/protocol_api/module_contexts.py:
```diff
@@ -148,6 +148,10 @@ def load_labware(  # noqa: C901
             The *label,* *namespace,* and *version* parameters.
+
+        .. versionadded:: 2.25
+            The *adapter_namespace*, *adapter_version,*
```
Oh, wild, I would use monospace here (double backticks, because sphinx) but ⭐ for following precedent from the above line. Change neither or change both.

SyntaxColoring replied:
Thanks for the validation that these ought to be monospace. Let's change both :shipit:

ecormany commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -459,18 +459,52 @@ def load_labware(  # noqa: C901
+        :param adapter_namespace: The namespace of the adapter being loaded.
+            Applies to ``adapter`` the same way that ``namespace`` applies to ``load_name``.
+
+            .. versionchanged:: 2.25
```
Needs to be bumped to 2.26 or wherever this lands.

I think we've done a mix of `versionadded` embedded within a param definition and `versionchanged` at the end of the entire ref entry. I generally prefer the latter. And wouldn't use `versionchanged` within a param unless it existed previously.

SyntaxColoring replied:
Bumped to 2.26.

Yeah, I hear you about `versionchanged` being weird if the param didn't exist previously. I tried a mixture of `versionadded`+`versionchanged` and it felt pretty noisy to me. I'll defer to you, but I think we'll leave those refinements to a separate PR.

ecormany reviewed (commented):
Docstrings are perfectly cromulent. Will leave open our RTC ticket in case we want to polish them up more before releasing this feature, but that shouldn't block any software-motivated launch timeline.

SyntaxColoring committed:
Extract repeated APIVersion(2, 25) to a constant.

SyntaxColoring committed:
2.25 -> 2.26.

SyntaxColoring committed:
Comma.

SyntaxColoring committed:
Italic -> monospace.

SyntaxColoring committed:
Parametrize ProtocolContext.load_labware() over multiple API versions.

SyntaxColoring committed:
Bump MAX_SUPPORTED_VERSION from 2.25 to 2.26 so it's covered in the tests.

SyntaxColoring committed:
Update test_load_labware_params.py test case for new APIVersion.

A new version of this labware got added in a separate PR, in an APIVersion that didn't exist at that time. This PR bumps the maximum APIVersion so it does exist, which caused the test to fail.

SyntaxColoring committed:
Merge chore_release-8.6.0@2d2d8f into version_and_namespace_args

Resolve docstring conflicts in api/src/opentrons/protocol_api/module_contexts.py. New parameter descriptions come from version_and_namespace_args. Spacing fixes come from version_and_namespace_args. Other content changes come from chore_release-8.6.0.

ddcc4 reviewed (approved):
Thank you. Please put this into the `hold-for-8.6.1` branch instead of `edge`. Once we cut the official `chore_release-8.6.1` branch, I'll merge this change (plus all the other ones in `hold-for-8.6.1`) into RS 8.6.1.

SyntaxColoring committed:
Delete leftover todo comment from development.

CaseyBatten commented on api/src/opentrons/protocols/api_support/definitions.py:
```diff
@@ -1,6 +1,6 @@
 from .types import APIVersion
 
-MAX_SUPPORTED_VERSION = APIVersion(2, 25)
+MAX_SUPPORTED_VERSION = APIVersion(2, 26)
```
Do we want this global API version maximum changed in this PR, or a different one related to the API 2.26 release? Any other API changes we think will happen in the next major release that would be bundled under API 2.26 that this would open up access to?(8.6.1? 8.7?) Doesn't matter too much given where this is merging, just something to take note of.

SyntaxColoring replied:
Yeah, I'm not really sure.

One practical effect of doing it now is that all of the `opentrons.protocol_api` unit tests that parametrize via `versions_above()` will now automatically run on `APIVersion(2, 26)`, which *is* what we want. If we did the 2.25->2.26 bump later, all of the associated test breakage would fall on whoever happened to do the bump.

ddcc4 replied:
Please do it now in this PR.

This PR will go into the `hold-for-8.6.1` branch, which is the branch that will introduce the new API version 2.26.

CaseyBatten reviewed (approved):
Looks really good, should make handling these labware group loads more sensible going forward. Thanks for doing this Max.

sfoster1 reviewed (approved):
Looks good to me, thanks!
